### DISCUSSION TITLE
Some questions related to seaweedfs architecture (What is volume and where files offsets are stored)

### DISCUSSION THREAD

**Discussion by strogiyotec** [seaweedfs/seaweedfs#2425] [Q&A]
Sponsors SeaweedFS via Patreon https://www.patreon.com/seaweedfs

**Describe the bug**
Good day, I want to try seaweedfs for our production server that stores images. After reading Haystack paper and wiki of seaweedfs I have few questions related to architecture. I hope I can get answers. Thanks in advance 
1. So according to wiki there are two concepts that represent a volume. Volume service(an http server) and just Volume(single actual file). When master generates a `fid` the first number before comma is volume id. Is it id of a file or id of a volume server ? If it's id of a file then each volume server has unique set of file ids ? If so how uniqueness is guaranteed ? 
2. In the example from READ.me with fid 3,01637037d6,  this part `637037d6` is a cookie. So let's say I have a single volume file and I want to read a photo from this file, as far as I understand seaweedfs has some in-memory storage(because in Haystack paper it's talking about storing this information in memory) where `637037d6` is a key and value is an offset and length of this photo in a volume file. If so then which memory storage is used to keep offsets ? How this storage behave if server will be restarted ?

I appreciate your time. After getting an answer I can contribute to seaweedfs by improving documentation for new users with similar questions.  Regards!

**Comment by chrislusf** [comment]
1. The `fid` format is <volume_id>, <file_key><cookie>. The file_key is generated by master in memory incrementally. The max file key is persisted in a local directory and also collected from existing volumes. It is also optional to use snowflake to generate file keys.
2. the file key is `01` in `3,01637037d6`. The memory storage is a needle map. If restarted, the needle map will be loaded from the volume's idx file.
